File: tests/compile.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { compile } from "../src/compile.js";

test("classifies root as static", () => {
  const compiled = compile({ "/": "index" });
  assert.equal(compiled.staticRoutes.get("/").value, "index");
  assert.equal(compiled.dynamicRoutes.length, 0);
});

test("classifies static routes", () => {
  const compiled = compile({ "/articles": "a", "/about": "b" });
  assert.equal(compiled.staticRoutes.size, 2);
  assert.equal(compiled.staticRoutes.get("/about").value, "b");
});

test("classifies dynamic routes", () => {
  const compiled = compile({ "/articles/:slug": "a" });
  assert.equal(compiled.staticRoutes.size, 0);
  assert.equal(compiled.dynamicRoutes.length, 1);
  assert.deepEqual(compiled.dynamicRoutes[0].paramNames, ["slug"]);
});

test("classifies wildcard", () => {
  const compiled = compile({ "*": "fallback" });
  assert.equal(compiled.wildcard.value, "fallback");
});

test("classifies mixed collection", () => {
  const compiled = compile({
    "/": "home",
    "/articles/:slug": "dynamic",
    "*": "fallback",
  });
  assert.equal(compiled.staticRoutes.size, 1);
  assert.equal(compiled.dynamicRoutes.length, 1);
  assert.equal(compiled.wildcard.value, "fallback");
});

File: go/compile_test.go
package roution

import (
	"reflect"
	"testing"
)

func TestCompileClassifiesRoot(t *testing.T) {
	compiled := Compile(map[string]any{"/": "home"})
	if got := compiled.StaticRoutes["/"].Value; got != "home" {
		t.Fatalf("static root value = %v, want home", got)
	}
	if len(compiled.DynamicRoutes) != 0 {
		t.Fatalf("dynamic routes = %d, want 0", len(compiled.DynamicRoutes))
	}
}

func TestCompileClassifiesStatic(t *testing.T) {
	compiled := Compile(map[string]any{"/articles": "a", "/about": "b"})
	if len(compiled.StaticRoutes) != 2 {
		t.Fatalf("static routes = %d, want 2", len(compiled.StaticRoutes))
	}
	if got := compiled.StaticRoutes["/about"].Value; got != "b" {
		t.Fatalf("about value = %v, want b", got)
	}
}

func TestCompileClassifiesDynamic(t *testing.T) {
	compiled := Compile(map[string]any{"/articles/:slug": "a"})
	if len(compiled.StaticRoutes) != 0 {
		t.Fatalf("static routes = %d, want 0", len(compiled.StaticRoutes))
	}
	if len(compiled.DynamicRoutes) != 1 {
		t.Fatalf("dynamic routes = %d, want 1", len(compiled.DynamicRoutes))
	}
	if got := compiled.DynamicRoutes[0].ParamNames; !reflect.DeepEqual(got, []string{"slug"}) {
		t.Fatalf("param names = %v, want [slug]", got)
	}
}

func TestCompileClassifiesWildcard(t *testing.T) {
	compiled := Compile(map[string]any{"*": "fallback"})
	if compiled.Wildcard == nil || compiled.Wildcard.Value != "fallback" {
		t.Fatalf("wildcard = %v, want fallback", compiled.Wildcard)
	}
}

func TestCompileMixedCollection(t *testing.T) {
	compiled := Compile(map[string]any{
		"/":            "home",
		"/articles/:slug": "dynamic",
		"*":            "fallback",
	})
	if len(compiled.StaticRoutes) != 1 {
		t.Fatalf("static routes = %d, want 1", len(compiled.StaticRoutes))
	}
	if len(compiled.DynamicRoutes) != 1 {
		t.Fatalf("dynamic routes = %d, want 1", len(compiled.DynamicRoutes))
	}
	if compiled.Wildcard == nil || compiled.Wildcard.Value != "fallback" {
		t.Fatalf("wildcard = %v, want fallback", compiled.Wildcard)
	}
}
